src/radiation/raytrace/raytrace.hpp

```diff
@@ -0,0 +1,285 @@
+}
+
+//----------------------------------------------------------------------------------------
+//! \fn  StateDescriptor RT::PushParticlesImpl
+//! \brief Implementation for pushing particles
+template <Coordinates GEOM, bool LOGR>
+TaskStatus PushParticlesImpl(MeshData<Real> *md) {
+  auto pm = md->GetParentPointer();
+  auto &resolved_pkgs = pm->resolved_packages;
+  auto &rt_pkg = pm->packages.Get("raytrace");
+  const auto efloor = rt_pkg->template Param<Real>("efloor");
+  const auto x1max = rt_pkg->template Param<Real>("x1max");
+  const auto x1min = rt_pkg->template Param<Real>("x1min");
+  const auto rstar = rt_pkg->template Param<Real>("stellar_radius");
+  // Create SparsePack
+  static auto desc =
+      MakePackDescriptor<rad::opac::cross_section, gas::src::energy>(resolved_pkgs.get());
+  auto vmesh = desc.GetPack(md);
+
+  // Create SwarmPacks
+  static auto pdesc_r =
+      MakeSwarmPackDescriptor<swarm_position::x, swarm_position::y, swarm_position::z,
+                              rad::part::flux>("star");
+  static auto pdesc_i = MakeSwarmPackDescriptor<rad::part::ijk>("star");
+  auto ppack_r = pdesc_r.GetPack(md);
+  auto ppack_i = pdesc_i.GetPack(md);
+
+  // Indexing and dimensionality
+  const int ndim = pm->ndim;
+  const bool multi_d = (ndim >= 2);
+  const bool three_d = (ndim == 3);
+  const int &nblocks = vmesh.GetNBlocks();
+  const int &nparticles_per_pack = ppack_r.GetMaxFlatIndex();
+  const auto ib = md->GetBoundsI(IndexDomain::interior);
+  const auto jb = md->GetBoundsJ(IndexDomain::interior);
+  const auto kb = md->GetBoundsK(IndexDomain::interior);
+
+  const int ngh = parthenon::Globals::nghost;
+
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "TransportPhotons", DevExecSpace(), 0, nparticles_per_pack,
+      KOKKOS_LAMBDA(const int idx) {
+        auto [b, n] = ppack_r.GetBlockParticleIndices(idx);
+        const auto &swarm_d = ppack_r.GetContext(b);
+        if (swarm_d.IsActive(n)) {
+          Real &ee = ppack_r(b, rad::part::flux(), n);
+          Real &xp = ppack_r(b, swarm_position::x(), n);
+          const Real &yp = ppack_r(b, swarm_position::y(), n);
+          const Real &zp = ppack_r(b, swarm_position::z(), n);
+          int &i = ppack_i(b, rad::part::ijk(0), n);
+          int &j = ppack_i(b, rad::part::ijk(1), n);
+          int &k = ppack_i(b, rad::part::ijk(2), n);
+          const auto &pco = vmesh.GetCoordinates(b);
+          const auto inds = GetIndices(pco, {xp, yp, zp});
+          i = ib.s + inds[0] - ngh;
+          if (multi_d) j = jb.s + inds[1] - ngh;
+          if (three_d) k = kb.s + inds[2] - ngh;
+
+          while ((i <= ib.e) && (ee > 0.0)) {
+            geometry::Coords<GEOM> coords(LOGR, pco, k, j, i);
+
+            // Deposit energy for this cell and decrement the photon energy
+            const auto dx = coords.bnds.x1[1] - coords.bnds.x1[0];
+            Real dtau = vmesh(b, rad::opac::cross_section(), k, j, i);
+            if (xp <= x1min + 1e-10) {
```

Comment:
This block is checking if the particle is at the lowest radial mesh boundary, then attenuating the particle weight from 6 times the stellar radius to the lowest boundary radius? Is this then assuming that 6 stellar radii is where the gas starts (i.e. the disk is evacuated below 6 * rstar? Can/should the 6 be set to a user parameter?